src/devices/MatrixKeyboard/MatrixKeyboard.cs

```diff
@@ -0,0 +1,231 @@
+    {
+        /// <summary>
+        /// Get row pins
+        /// </summary>
+        public IEnumerable<int> RowPins => _rowPins;
+
+        /// <summary>
+        /// Get column pins
+        /// </summary>
+        public IEnumerable<int> ColumnPins => _columnPins;
+
+        /// <summary>
+        /// Get or set scanning frequency in Hertz
+        /// </summary>
+        public double ScanFrequency
+        {
+            get => _scanFrequency;
+            set
+            {
+                if (value <= 0)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), "Scanning frequency must be positive");
+                }
+                _scanFrequency = value;
+                _scanInterval = CalcScanInterval(_scanFrequency);
+            }
+        }
+
+        /// <summary>
+        /// Get whether this keyboard is scanning
+        /// </summary>
+        public bool IsScanning { get; private set; }
+
+        private readonly int[] _rowPins;
+        private readonly int[] _columnPins;
+        private int _currentRow;
+        private readonly PinValue[] _buttonValues;
+        private double _scanFrequency;
+
+        /// <summary>
+        /// Initialize keyboard
+        /// </summary>
+        /// <param name="gpioController">GPIO controller</param>
+        /// <param name="rowPins">Row pins</param>
+        /// <param name="colPins">Column pins</param>
+        /// <param name="scanFreq">Scanning frequency</param>
+        public MatrixKeyboard(IGpioController gpioController, IEnumerable<int> rowPins, IEnumerable<int> colPins, double scanFreq = 50)
+        {
+            _gpio = gpioController ?? throw new ArgumentNullException(nameof(gpioController));
+
+            if (rowPins == null)
+            {
+                throw new ArgumentNullException(nameof(rowPins));
+            }
+            if (rowPins.Count() < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(rowPins), "The number of rows is at least 1");
+            }
+
+            if (colPins == null)
+            {
+                throw new ArgumentNullException(nameof(colPins));
+            }
+            if (colPins.Count() < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(colPins), "The number of columns is at least 1");
+            }
+
+            if (scanFreq <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scanFreq), "Scanning frequency must be positive");
+            }
+
+            _rowPins = rowPins.ToArray();
+            _columnPins = colPins.ToArray();
+            _buttonValues = new PinValue[_rowPins.Length * _columnPins.Length];
+
+            _scanThread = new Thread(ScanThreadStart);
+            IsScanning = false;
+            ScanFrequency = scanFreq;
+        }
+
+        /// <summary>
+        /// Open GPIO pins then start keyboard scanning
+        /// </summary>
+        public void StartScan()
```

Comment:
This should be:
`Task Scan(CancellationToken cancellationToken)`

This way you can avoid most of the issues mentioned here (still might be worth preventing from having multiple loops).